RFC-0004-functions-stats.md

```diff
@@ -0,0 +1,473 @@
+Since we are introducing fields inside the
+[FunctionMetadata.h](https://github.com/facebookincubator/velox/blob/0adc62e3aaf647596ab476f6300a4165e60f6b91/velox/expression/FunctionMetadata.h)
+the new fields have to be sent as is for further processing to the coordinator.
+
+In both Java and C++, builtin functions have extra metadata defined in FunctionMetadata class, which is used while computing estimated stats
+for optimizer.
+
+### Phase 3: Annotate some of the builtin scalar functions with appropriate annotations.
+
+See: [Appendix](#appendix) 1, for list of functions.
+
+### Phase 4: Extend the support to non-built-in UDFs. Add documentation and blogs with usage examples.
+
+### Future work:
+
+#### Define an expression language:
+The following example is a more powerful form of expressing how to estimate stats for functions based on their characteristics. Here,
+we propose the use of expressions ( i.e. an expression language with a fixed grammar ) In the following example of `substr` :
+For average row size we have used the value of length argument. For distinct values count it is more complex, e.g. if length is same as
+the upper bound for `varchar(x)` i.e. `x` , then use source stat `distinct values count` of the argument `utf8`. 
+```java
+    @Description("substring of given length starting at an index")
+    @ScalarFunction
+    @LiteralParameters("x")
+    @SqlType("varchar(x)")
+    public static Slice substr(
+            @ScalarPropagateSourceStats(
+                    nullFraction = USE_SOURCE_STATS,
+                    distinctValuesCount = "value(arg_length) <3 ? estimate_ndv(x, value(arg_length), BASIC_LATIN, min(output_row_count, source_stats_ndv())) * (1- source_stats_null_fraction()) : UNKNOWN",
+                    avgRowSize = "value(arg_length)" ) @SqlType("varchar(x)") Slice utf8,
+            @SqlType(StandardTypes.BIGINT) long start,
+            @SqlType(StandardTypes.BIGINT) long length)
+    { //...
+    }
+```
+
+`substr` is at the heart of all `LIKE %X` statements, which is widely used. Without a more powerful framework support it is difficult to support a stats estimation. 
+
+#### virtual function
+C++ [VectorFunction](https://github.com/facebookincubator/velox/blob/0adc62e3aaf647596ab476f6300a4165e60f6b91/velox/expression/VectorFunction.h) class definition to include
+virtual function which can be overridden by each function to provide how the source stats are transformed.
+
+## Metrics
+
+How can we measure the impact of this feature?
+
+1. Existing TPCDS and TPCH queries which use functions or `LIKE %STR` in join condition will benefit from this optimization.
+2. Join order benchmark: Reference [2](#references)
+
+## Other Approaches Considered
+
+Based on the discussion, this may need to be updated with feedback from reviewers.
+
+### Approach 1 Trying to find a heuristic based mechanism to estimate functions cost, for example : A function takes a single argument of type `varchar(20)` and
```

Comment:
This and `Approach 2` are very long headings. Did you mean to add a line break after `Approach 1` ?